cmd/build/v2/services.go

```diff
@@ -106,23 +108,93 @@ func (bc *OktetoBuilder) checkServicesToBuild(service string, manifest *model.Ma
+			continue
+		} else if err != nil {
+			// return error if the registry doesn't send a not found error
+			return "", fmt.Errorf("error checking image at registry %s: %v", tag, err)
+		}
+		return imageWithDigest, nil
+	}
+	return "", fmt.Errorf("not found")
+}
+
+func (bc *OktetoBuilder) tagsToCheck(manifestName, svcName string, b *model.BuildInfo) []string {
+	targetRegistries := []string{constants.DevRegistry}
+	sha := ""
+	if bc.Config.HasGlobalAccess() && bc.Config.IsCleanProject() {
+		targetRegistries = []string{constants.GlobalRegistry, constants.DevRegistry}
+		sha = bc.Config.GetHash()
+	}
+
 	// manifestName can be not sanitized when option name is used at deploy
 	sanitizedName := format.ResourceK8sMetaString(manifestName)
+	tagsToCheck := []string{}
+
 	switch {
 	case !okteto.IsOkteto():
-		return b.Image
+		tagsToCheck = append(tagsToCheck, b.Image)
+		return tagsToCheck
 	case (shouldBuildFromDockerfile(b) && shouldAddVolumeMounts(b)) || shouldAddVolumeMounts(b):
-		return fmt.Sprintf("%s/%s-%s:%s", targetRegistry, sanitizedName, svcName, model.OktetoImageTagWithVolumes)
+		if sha != "" {
+			for _, targetRegistry := range targetRegistries {
+				tagsToCheck = append(tagsToCheck, getImageFromTmpl(targetRegistry, sanitizedName, svcName, sha))
+			}
+		}
+		for _, targetRegistry := range targetRegistries {
+			tagsToCheck = append(tagsToCheck, getImageFromTmpl(targetRegistry, sanitizedName, svcName, model.OktetoImageTagWithVolumes))
+		}
+		return tagsToCheck
 	case b.Image != "" && shouldBuildFromDockerfile(b):
-		return b.Image
+		tagsToCheck = append(tagsToCheck, b.Image)
+		return tagsToCheck
 	case shouldBuildFromDockerfile(b):
-		return fmt.Sprintf("%s/%s-%s:%s", targetRegistry, sanitizedName, svcName, model.OktetoDefaultImageTag)
+		if sha != "" {
+			for _, targetRegistry := range targetRegistries {
+				tagsToCheck = append(tagsToCheck, getImageFromTmpl(targetRegistry, sanitizedName, svcName, sha))
+			}
+		}
+		for _, targetRegistry := range targetRegistries {
+			tagsToCheck = append(tagsToCheck, getImageFromTmpl(targetRegistry, sanitizedName, svcName, model.OktetoDefaultImageTag))
+		}
+		return tagsToCheck
 	case b.Image != "":
-		return b.Image
+		tagsToCheck = append(tagsToCheck, b.Image)
+		return tagsToCheck
 	default:
 		oktetoLog.Infof("could not build service %s, due to not having Dockerfile defined or volumes to include", svcName)
 	}
-	return ""
+	return tagsToCheck
+}
+
+func getImageFromTmpl(targetRegistry, repoName, svcName, tag string) string {
+	return fmt.Sprintf("%s/%s-%s:%s", targetRegistry, repoName, svcName, tag)
+}
+
+func (bc *OktetoBuilder) checkIfCommitIsAlreadyBuilt(ctx context.Context, manifestName, svcName string, options *types.BuildOptions) (string, bool) {
+	if !bc.Config.IsCleanProject() {
+		return "", false
+	}
+	if options.NoCache {
+		return "", false
+	}
+	sha := bc.Config.GetHash()
+
+	targetRegistries := []string{constants.GlobalRegistry, constants.DevRegistry}
```

Comment:
at this point, could we known the target registry to only check it there if image is already built?